### PR TITLE
[wip] superset-ui-plugins linking tool for local development

### PR DESCRIPTION
**Pull request by suddjian**
### CATEGORY

Choose one

- [ ] Bug Fix
- [ ] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [x] Build / Development Environment
- [ ] Documentation

### SUMMARY

This adds a tool to aid local plugin development/debugging by linking all of the packages in the `superset-ui-plugins` repo. A new npm script, `plugin-devmode-on`, `npm link`s all the superset-ui-plugins. A new webpack config detects any npm linked superset-ui-plugins, and automatically changes imports for those plugins to import from `/src`. This makes a [whole series of steps](https://github.com/apache-superset/superset-ui/blob/master/docs/debugging.md) unnecessary. You can set the location to use for the repo with the env variable `SUPERSET_UI_PLUGINS_PATH`.

Another npm script `plugin-devmode-off` undoes the changes of `plugin-devmode-on` by running `npm ci`.

*This is not entirely working, just yet.* The `BoxPlotChartPlugin` in `@superset-ui/preset-chart-xy` plugin is imported by reaching into the `/esm/legacy` directory. This breaks, because webpack rewrites it to look for `src/esm/legacy`, which does not exist. We propose to solve this by either modernizing the plugin (no idea how easy/difficult that is), or splitting it into a legacy version and a non-legacy version, so that we don't need to reach inside a build folder to import the chart. Input would be appreciated from anyone with more insight into the history of that plugin.

### CODE REVIEW

**Comment by etr2460** [comment]
@williaster mind taking a look at this in lieu of @kristw since he's on leave?

**Review by mistercrunch — CHANGES_REQUESTED** [review]
README docs as to how to use this should be in scope for this PR

File: superset/assets/plugin-devmode.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * under the License.
+ */
+
+const fs = require('fs');
+const path = require('path');
+const { execSync } = require('child_process');
+
+const PLUGINS_REPO =
+  process.env.SUPERSET_UI_PLUGINS_PATH ||
+  path.resolve('../../../superset-ui-plugins');
```
Comment (mistercrunch) [inline superset/assets/plugin-devmode.js]: future: eventually we'll have to add support for other repos like https://github.com/apache-superset/superset-ui-plugins-deckgl

**Comment by suddjian** [comment]
There's an issue with the approach used here: The webpack aliasing is checking for any linked superset-ui-plugins packages, and modifying the imports to import from `/src` automatically, but this makes it effectively impossible to debug problems that only arise after build time.

This can be fixed by introducing a `plugin-devmode.json` file, which will store the data about which plugins are in devmode instead of using the status of symlinked packages to get that info.

**Comment by ktmud** [comment]
Was anyone tempted to move `superset-ui` (especially the visualization registry) and `superset-ui-plugins` (or at least the basic ones) back to the core frontend code? I'm struggling to see the benefit of putting them in separate repos. They are core to Superset's basic functions, not really pluggable, and unlikely to be used by other projects. 

Currently we haven't migrated off the legacy plugins and  the added overhead of syncing package versions, maintaining separate building processes only makes the migration/iterations slower.

Both [Redash](https://github.com/getredash/redash/tree/master/client/app/visualizations) and [Metabase](https://github.com/metabase/metabase/tree/master/frontend/src/metabase/visualizations) didn't take this approach. IMO their visualization registries look much cleaner than Superset and are more developer-friendly to new contributors.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -805,6 +805,12 @@ Here's an example as a Github PR with comments that describe what the
 different sections of the code do:
 https://github.com/apache/incubator-superset/pull/3013
 
+To run Superset using visualization plugins from your local machine, run `npm run plugin-devmode-on`. This will use `npm link` to connect the Superset frontend to your local copy of [superset-ui-plugins](https://github.com/apache-superset/superset-ui-plugins). This command uses the environment variable `SUPERSET_UI_PLUGINS_PATH` to determine the path, defaulting to `../superset-ui-plugins`. So, for examplem, if the path to your local `superset-ui-plugins` repo, relative to `incubator-superset`, happens to be `../../my-fork-of-superset-ui-plugins`, then run `SUPERSET_UI_PLUGINS_PATH=../../my-fork-of-superset-ui-plugins npm run plugin-devmode-on`
```
Comment (villebro) [inline CONTRIBUTING.md]: typo: "examplem"

**Comment by mistercrunch** [comment]
@ktmud I agree that it's been really hard to work on plugins for quite some time now. I'm really hoping we can make it much easier to work with plugins quickly.

I've been wondering whether it could make sense to bring back some of the core plugins to the main repo too. The long tail and the community-contributed ones could live in other repos, but having the core in the main repo/package may help. PRing on two repos, publishing packages and bumping versions is tricky, and that's on top of the `npm link` stuff.

It's been many times that I want to work in a plugin and either really struggle or completely fail at it.

**Comment by rusackas** [comment]
Closing in favor of https://github.com/apache/incubator-superset/pull/9326

**Pull request closed without merging**